File: android/feature/measurements/src/main/kotlin/xyz/klinik/feature/measurements/DecimalEntry.kt
package xyz.klinik.feature.measurements

/**
 * Reading a number the way a person typed it.
 *
 * A Turkish keyboard produces "72,4" and an English one "72.4", and both reach
 * this app — patients arrive from many countries and the clinic's own staff
 * type in Turkish. `"72,4".toDoubleOrNull()` is null, so a naive parse turns a
 * perfectly good weight into a blank field.
 *
 * Deliberately locale-independent: it accepts either separator rather than the
 * one the device is set to, because the device locale and the keyboard the
 * person is using are not the same thing.
 */
object DecimalEntry {
    fun parse(text: String): Double? {
        val trimmed = text.trim()
        if (trimmed.isEmpty()) return null

        // Anything that is not digits and a single separator is refused rather
        // than parsed into something plausible: a weight drives dosing.
        if (!trimmed.all { it.isDigit() || it == '.' || it == ',' }) return null
        if (trimmed.count { it == '.' || it == ',' } > 1) return null

        val normalised = trimmed.replace(',', '.')
        if (normalised.endsWith(".")) return null

        return normalised.toDoubleOrNull()
    }
}

File: ios/Sources/KlinikMeasurementsFeature/DecimalEntry.swift
import Foundation

/// Reading a number the way a person typed it.
///
/// A Turkish keyboard produces "72,4" and an English one "72.4", and both reach
/// this app — patients arrive from many countries and the clinic's own staff
/// type in Turkish. `Double("72,4")` is nil, so a naive parse silently turns a
/// perfectly good weight into a blank field, or worse, into 72.
///
/// Deliberately locale-independent: it accepts either separator rather than
/// the one the device is set to, because the device locale and the keyboard
/// the person is using are not the same thing.
public enum DecimalEntry {
    public static func parse(_ text: String) -> Double? {
        let trimmed = text.trimmingCharacters(in: .whitespaces)
        guard !trimmed.isEmpty else { return nil }

        // Reject anything that is not digits and a single separator, so "72.4.5"
        // or "12kg" are refused rather than parsed to something plausible.
        let allowed = CharacterSet(charactersIn: "0123456789.,")
        guard trimmed.unicodeScalars.allSatisfy({ allowed.contains($0) }) else { return nil }

        let separators = trimmed.filter { $0 == "." || $0 == "," }
        guard separators.count <= 1 else { return nil }

        // Whichever separator the person used, it means the same thing here:
        // these are single measurements, never thousands-grouped.
        let normalised = trimmed.replacingOccurrences(of: ",", with: ".")
        guard !normalised.hasSuffix(".") else { return nil }

        // Parsed in a fixed locale, because by this point the text has been
        // normalised to one form; using the device locale here would undo that.
        return Double(normalised)
    }
}
